File: frontend/src/api/voice.ts
/**
 * API client functions for the voice discussion mode.
 */

import { DEFAULT_PROFILE_ID } from "./applications";

const API_BASE = "/api/voice";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface VoiceMessage {
  id: number;
  session_id: number;
  role: "user" | "assistant";
  content: string;
  created_at: string;
}

export interface VoiceSession {
  id: number;
  profile_id: number;
  application_id: number | null;
  mode: "cover_letter" | "coaching" | "job_evaluation";
  title: string | null;
  status: "active" | "completed";
  messages: VoiceMessage[];
  created_at: string;
  updated_at: string;
}

export interface VoiceSessionListResponse {
  sessions: VoiceSession[];
  total: number;
}

export interface VoiceSendResponse {
  user_message: VoiceMessage;
  assistant_message: VoiceMessage;
  session: VoiceSession;
}

export type VoiceMode = "cover_letter" | "coaching" | "job_evaluation";

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function createVoiceSession(params: {
  mode: VoiceMode;
  application_id?: number;
  title?: string;
}): Promise<VoiceSession> {
  const res = await fetch(`${API_BASE}/sessions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      profile_id: DEFAULT_PROFILE_ID,
      ...params,
    }),
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(
      (body as { detail?: string }).detail ??
        `Failed to create voice session: ${res.status}`,
    );
  }
  return res.json() as Promise<VoiceSession>;
}

export async function fetchVoiceSessions(
  mode?: VoiceMode,
): Promise<VoiceSessionListResponse> {
  const params = new URLSearchParams({
    profile_id: String(DEFAULT_PROFILE_ID),
  });
  if (mode) params.set("mode", mode);

  const res = await fetch(`${API_BASE}/sessions?${params.toString()}`);
  if (!res.ok) {
    throw new Error(`Failed to fetch voice sessions: ${res.status}`);
  }
  return res.json() as Promise<VoiceSessionListResponse>;
}

export async function fetchVoiceSession(
  sessionId: number,
): Promise<VoiceSession> {
  const res = await fetch(
    `${API_BASE}/sessions/${sessionId}?profile_id=${DEFAULT_PROFILE_ID}`,
  );
  if (!res.ok) {
    throw new Error(`Failed to fetch voice session: ${res.status}`);
  }
  return res.json() as Promise<VoiceSession>;
}

export async function sendVoiceMessage(
  sessionId: number,
  content: string,
): Promise<VoiceSendResponse> {
  const res = await fetch(`${API_BASE}/sessions/${sessionId}/messages`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      profile_id: DEFAULT_PROFILE_ID,
      content,
    }),
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(
      (body as { detail?: string }).detail ??
        `Failed to send voice message: ${res.status}`,
    );
  }
  return res.json() as Promise<VoiceSendResponse>;
}

export async function completeVoiceSession(
  sessionId: number,
): Promise<VoiceSession> {
  const res = await fetch(
    `${API_BASE}/sessions/${sessionId}/complete?profile_id=${DEFAULT_PROFILE_ID}`,
    { method: "POST" },
  );
  if (!res.ok) {
    throw new Error(`Failed to complete voice session: ${res.status}`);
  }
  return res.json() as Promise<VoiceSession>;
}

File: src/career_os/api/voice.py
"""Voice Discussion Mode API routes."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from career_os.api.constants import DESC_ACTIVE_PROFILE_ID, RESP_404, RESP_404_422
from career_os.database import get_db
from career_os.schemas.voice import (
    VoiceMessageCreate,
    VoiceMessageResponse,
    VoiceSendResponse,
    VoiceSessionCreate,
    VoiceSessionListResponse,
    VoiceSessionResponse,
)
from career_os.services.voice import (
    ApplicationNotFoundError,
    InvalidModeError,
    ProfileNotFoundError,
    SessionNotFoundError,
    complete_session,
    create_session,
    get_session,
    list_sessions,
    send_message,
)

router = APIRouter(prefix="/api/voice", tags=["voice"])


@router.post(
    "/sessions",
    status_code=201,
    responses=RESP_404_422,
)
async def create_voice_session(
    data: VoiceSessionCreate,
    db: Annotated[Session, Depends(get_db)],
) -> VoiceSessionResponse:
    """Create a new voice discussion session.

    Modes:
    - cover_letter: brainstorm cover letter (requires application_id)
    - coaching: career coaching dialogue
    - job_evaluation: evaluate a job opportunity (requires application_id)
    """
    try:
        session = create_session(
            db,
            profile_id=data.profile_id,
            mode=data.mode,
            application_id=data.application_id,
            title=data.title,
        )
    except ProfileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ApplicationNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except InvalidModeError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc

    return _session_response(session)


@router.get("/sessions")
async def list_voice_sessions(
    profile_id: Annotated[int, Query(description=DESC_ACTIVE_PROFILE_ID)],
    db: Annotated[Session, Depends(get_db)],
    mode: Annotated[str | None, Query(description="Filter by mode")] = None,
) -> VoiceSessionListResponse:
    """List voice discussion sessions for a profile."""
    sessions = list_sessions(db, profile_id=profile_id, mode=mode)
    return VoiceSessionListResponse(
        sessions=[_session_response(s) for s in sessions],
        total=len(sessions),
    )


@router.get("/sessions/{session_id}", responses=RESP_404)
async def get_voice_session(
    session_id: int,
    profile_id: Annotated[int, Query(description=DESC_ACTIVE_PROFILE_ID)],
    db: Annotated[Session, Depends(get_db)],
) -> VoiceSessionResponse:
    """Get a voice session with all messages."""
    try:
        session = get_session(db, session_id=session_id, profile_id=profile_id)
    except SessionNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

    return _session_response(session)


@router.post(
    "/sessions/{session_id}/messages",
    responses=RESP_404,
)
async def send_voice_message(
    session_id: int,
    data: VoiceMessageCreate,
    db: Annotated[Session, Depends(get_db)],
) -> VoiceSendResponse:
    """Send a message in a voice session and receive AI response.

    Accepts text from any STT tool (SuperWhisper, MacWhisper, system dictation)
    or directly typed input.
    """
    try:
        user_msg, assistant_msg = await send_message(
            db,
            session_id=session_id,
            profile_id=data.profile_id,
            content=data.content,
        )
    except SessionNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

    session = get_session(db, session_id=session_id, profile_id=data.profile_id)

    return VoiceSendResponse(
        user_message=VoiceMessageResponse.model_validate(user_msg),
        assistant_message=VoiceMessageResponse.model_validate(assistant_msg),
        session=_session_response(session),
    )


@router.post(
    "/sessions/{session_id}/complete",
    responses=RESP_404,
)
async def complete_voice_session(
    session_id: int,
    profile_id: Annotated[int, Query(description=DESC_ACTIVE_PROFILE_ID)],
    db: Annotated[Session, Depends(get_db)],
) -> VoiceSessionResponse:
    """Mark a voice session as completed."""
    try:
        session = complete_session(db, session_id=session_id, profile_id=profile_id)
    except SessionNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

    return _session_response(session)


def _session_response(session) -> VoiceSessionResponse:
    """Convert a VoiceSession ORM object to response schema."""
    return VoiceSessionResponse(
        id=session.id,
        profile_id=session.profile_id,
        application_id=session.application_id,
        mode=session.mode,
        title=session.title,
        status=session.status,
        messages=[VoiceMessageResponse.model_validate(m) for m in session.messages],
        created_at=session.created_at,
        updated_at=session.updated_at,
    )
